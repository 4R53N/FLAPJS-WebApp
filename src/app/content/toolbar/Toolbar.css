.toolbar-container {
  padding-left: 10px;
  display: flex;
  flex-direction: row;

  height: var(--toolbar-height);
  overflow: hidden;
  overflow-x: scroll;

  background-color: var(--color-toolbar-main);
  box-shadow: 0 0.16em 0 var(--color-toolbar-accent-dark);
}
.toolbar-container .icon-button {
  fill: var(--color-toolbar-text);
}

.toolbar-title {
  display: flex;
  flex-direction: column;

  width: 12em;
  min-width: 12em;
}
.toolbar-title-name {
  display: flex;
  flex-direction: row;
  width: 100%;

  border-bottom: 1px solid var(--color-toolbar-text-invert);
}
#machine-name {
  background: none;
  border: none;
  padding: 0;

  width: 100%;
  font-size: 2em;

  opacity: 0.3;
  color: var(--color-toolbar-text);
  transition: opacity 0.3s ease;
}
#machine-name:hover {
  opacity: 0.5;
}
#machine-name:focus {
  outline: none;

  opacity: 1;
}
#machine-type {
  cursor: pointer;
  margin-left: -5px;
  margin-right: -5px;
  padding-top: 2px;
  padding-bottom: 2px;
  padding-left: 5px;
  padding-right: 5px;
  font-size: 1em;

  color: var(--color-toolbar-accent);
  border-radius: 10px;

  background-color: rgba(255,255,255,0.1);
  transition: background-color 0.3s ease;
}
#machine-type:hover {
  background-color: rgba(255,255,255,0.2);
}
#toolbar-upload-alt {
  min-width: 3em;
  margin-top: auto;

  fill: var(--color-toolbar-main-dark);
  transition: fill 0.6s ease;
}
#toolbar-upload-alt svg {
  cursor: pointer;
  width: 100%;
  height: 100%;
}
#toolbar-upload-alt:hover, #machine-name:focus ~#toolbar-upload-alt {
  fill: white;
}

.toolbar-tray {
  display: flex;
  flex-direction: row;

  padding-left: 10px;
  width: calc(100% - 12em);/* full width - width of title */
}

.navicon {
  fill: white;
  justify-content: flex-end;

  margin-right: 5px;
  margin-bottom: 5px;
  width: 4em;
  height: 100%;
  min-width: 30px;

  transition: fill 0.3s ease;
}
.navicon svg {
  width: 100%;
  height: 100%;
}
.navicon:hover {
  fill: var(--color-toolbar-accent);
}
.navicon:disabled {
  fill: var(--color-toolbar-main-dark);
}

#toolbar-help {
  margin-left: auto;
}
#toolbar-help svg {
  width: 3em;
  height: 3em;
}
