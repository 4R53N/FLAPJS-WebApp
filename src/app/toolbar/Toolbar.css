.toolbar-container {
  display: flex;
  height: var(--toolbar-height);
  overflow: hidden;

  background: lightblue;
}
