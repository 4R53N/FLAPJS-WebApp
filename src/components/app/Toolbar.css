.toolbar {
    background: gold;
    width: 100%;
    height: 2rem;
}
