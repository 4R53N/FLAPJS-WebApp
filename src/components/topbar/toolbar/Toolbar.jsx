import React from 'react';
import PropTypes from 'prop-types';
// import Style from './Toolbar.module.css';

import ToolbarDropdown from '@flapjs/components/topbar/toolbar/dropdown/ToolbarDropdown.jsx';

/**
 * A React component that can do anything you want. :D.
 * Heavily inspired by https://blog.logrocket.com/building-a-custom-dropdown-menu-component-for-react-e94f02ced4a1/.
 */
class Toolbar extends React.Component
{
    constructor(props)
    {
        super(props);
        this.state = {
            file: [
                {
                    id: 0,
                    title: 'New',
                    selected: false,
                    key: 'file',
                    options: [
                        {
                            id: 3,
                            title: 'DFA',
                            selected: false,
                            key: 'file/new'
                        },
                        {
                            id: 4,
                            title: 'NFA',
                            selected: false,
                            key: 'file/new'
                        },
                        {
                            id: 5,
                            title: 'PDA',
                            selected: false,
                            key: 'file/new'
                        }
                    ]
                },
                {
                    id: 1,
                    title: 'Open',
                    selected: false,
                    key: 'file'
                },
                {
                    id: 2,
                    title: 'Save',
                    selected: false,
                    key: 'file'
                }
            ]
        };
        this.toggleSelected = this.toggleSelected.bind(this);
    }

    toggleSelected(id, key) 
    {
        let selection = this.state[key];
        selection[id].selected = !selection[id].selected;
        this.setState({
            [key]: selection
        });
    }

    /** @override */
    render()
    {
        const props = this.props;
        
        return (
            <div className={props.className}>
                <ToolbarDropdown
                    title="File"
                    list={this.state.file}
                    toggleItem={this.toggleSelected}
                />
            </div>
        );
    }
}

Toolbar.propTypes = {
    id: PropTypes.string,
    style: PropTypes.object,
    className: PropTypes.string,
};
Toolbar.defaultProps = {
};

export default Toolbar;
