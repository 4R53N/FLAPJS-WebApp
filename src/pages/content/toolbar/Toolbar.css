#machine-name {
  background: none;
  padding: 0;
  border: none;
  font-size: 30px;
  color: #B6B1A9;
  border-bottom: 1px solid #182B49;
  width: 6em;
  margin: 0 15px 0 0;
}
#machine-name:focus {
  outline: none;
}

#machine-type {
  color: #FFCD00;
  font-size: 25px;
  position: absolute;
  top: 35px;
  left: 10px;
  font-family: sans-serif;
}

.toolbar-container {
  padding-left: 10px;
  display: flex;

  height: var(--toolbar-height);
  overflow: hidden;

  background-color: #006A96;
  box-shadow: 0 0.16em 0 #FFCD00;
}

.toolbar-button {
  background: none;/*Disable normal button style*/
  border: none;/*Disable normal button style*/
  outline: none;/*Disable normal button style*/
  padding: 0;/*Disable normal button style*/
  fill: white;
  justify-content: flex-end;
  margin-right: 15px;


  width: 3em;
  height: 100%;

  transition: fill 0.3s ease;
}
.toolbar-button:hover {
  fill: #FFCD00;
}
.toolbar-button:disabled {
  fill: #005C82;
}
.navicons {
  width: 100%;
  height: 100%;
}

#toolbar-help {
  margin-left: auto;

  fill: #FFCD00;
  width: 3em;
}
#toolbar-help:hover {
  fill: #C69214;
}

#import-file {
  opacity: 0;
  position: absolute;
  z-index: -1;
}
