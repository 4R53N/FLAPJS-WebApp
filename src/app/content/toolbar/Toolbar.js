import React from 'react';
import './Toolbar.css';

import Config from 'config.js';

import UploadButton from "./UploadButton.js";
import HelpButton from "./HelpButton.js";
import MachineName from './MachineName.js';

import IconButton from 'icons/IconButton.js';

import MoreIcon from 'icons/MoreIcon.js';
import UploadIcon from 'icons/UploadIcon.js';
import CreateIcon from 'icons/PageNewIcon.js';
//import CreateIcon from 'icons/CreateIcon.js';
import SaveIcon from 'icons/SaveIcon.js';
//import SaveIcon from 'icons/DiskSaveIcon.js';
import UndoIcon from 'icons/UndoIcon.js';
import RedoIcon from 'icons/RedoIcon.js';
import HelpIcon from 'icons/HelpIcon.js';
import LanguageIcon from 'icons/LanguageIcon.js';
import OfflineIcon from 'icons/OfflineIcon.js';
import OfflinePinIcon from 'icons/OfflinePinIcon.js';
import BugReportIcon from 'icons/BugReportIcon.js';

import Drawer from 'content/drawer/Drawer.js';

class Toolbar extends React.Component
{
  constructor(props)
  {
    super(props);

    this.state = {
      langOn: false
    };

    this.onMachineNameChange = this.onMachineNameChange.bind(this);
    this.onMachineNameKeyDown = this.onMachineNameKeyDown.bind(this);
    this.langOnClick = this.langOnClick.bind(this);
  }

  onMachineNameChange(e)
  {
    this.props.machineController.renameMachine(e.value);
  }

  onMachineNameKeyDown()
  {

  }

  onMachineNameBlur()
  {

  }

  langOnClick()
  {
    this.setState({langOn: !this.state.langOn});
  }

  render()
  {
    const app = this.props.app;
    const graphController = this.props.graphController;
    const machineController = this.props.machineController;
    const undoManager = this.props.undoManager;
    const drawer = this.props.drawer;

    const graph = graphController.getGraph();
    const machineBuilder = machineController.getMachineBuilder();

    const offline = navigator && navigator.onLine;

    return <div className="toolbar-container">
      <div className="toolbar-title">
        <div className="toolbar-title-name">
          {/*Machine Name*/}
          <MachineName id="machine-name" value={machineController.getMachineName()} onSubmit={this.onMachineNameChange}/>
          {/*Toolbar Alt. Title*/}
          <div className="toolbar-title-alt">
            {/*Offline Button*/}
            <IconButton id="toolbar-lang" title={I18N.toString("action.toolbar.lang")}
              onClick={() => {
                if ('serviceWorker' in navigator)
                {
                  if (navigator.serviceWorker.getRegistration("./app.html"))
                  {
                    alert(I18N.toString("message.offline.ready"));
                  }
                  else
                  {
                    alert(I18N.toString("message.offline.failed"));
                  }
                }
                else
                {
                  alert(I18N.toString("message.offline.nosupport"));
                }
              }} disabled={offline}>
              <OfflineIcon/>
            </IconButton>
            {/*Upload Button
            <div id="toolbar-upload-alt"
              title={I18N.toString("action.toolbar.uploadmachine")}>
              <MoreIcon/>
            </div>
            */}
          </div>
        </div>
        {/*Machine Type*/}
        <label id="machine-type"
          onClick={()=>drawer.setTab(0)}>
          {machineBuilder.getMachineType()}
        </label>
      </div>

      <div className="toolbar-tray">
        {/*New Button*/}
        <IconButton className="navicon" id="toolbar-new" title={I18N.toString("action.toolbar.newmachine")}
          onClick={() => {
            if (window.confirm(I18N.toString("alert.graph.clear")))
            {
              graph.clear();
              undoManager.clear();
              machineController.setMachineName(null);
            }
          }}>
          <CreateIcon/>
        </IconButton>
        {/*Upload Button*/}
        <UploadButton className="navicon" id="toolbar-upload" title={I18N.toString("action.toolbar.uploadmachine")}
          app={app}
          graphController={graphController}>
          <UploadIcon/>
        </UploadButton>
        {/*Undo Button*/}
        <IconButton className="navicon" id="toolbar-undo" title={I18N.toString("action.toolbar.undo")}
          disabled={!undoManager.canUndo()}
          onClick={()=>undoManager.undo()}>
          <UndoIcon/>
        </IconButton>
        {/*Redo Button*/}
        <IconButton className="navicon" id="toolbar-redo" title={I18N.toString("action.toolbar.redo")}
          disabled={!undoManager.canRedo()}
          onClick={()=>undoManager.redo()}>
          <RedoIcon/>
        </IconButton>
        {/*Save Button*/}
        <IconButton className="navicon" id="toolbar-save" title={I18N.toString("action.toolbar.save")}
          onClick={()=>drawer.setTab(Drawer.EXPORTING_PANEL_TAB_INDEX)} disabled={graph.isEmpty()}>
          <SaveIcon/>
        </IconButton>

        {/*Right Alt. Toolbar*/}
        <div className="toolbar-tray-alt">
          {/*Report a Bug Button*/}
          <IconButton className="navicon" id="toolbar-bug" title={I18N.toString("action.toolbar.bug")}>
            <BugReportIcon/>
          </IconButton>
          {/*Help Button*/}
          <HelpButton className="navicon" id="toolbar-help" title={I18N.toString("action.toolbar.help")}>
            <HelpIcon/>
          </HelpButton>
          {/*Language Button*/}
          <IconButton className="navicon" id="toolbar-lang" title={I18N.toString("action.toolbar.lang")}
            onClick={(e) => {document.activeElement.blur(); this.langOnClick();}} >
            <LanguageIcon/>
          </IconButton>
          {/*Language Select*/}
          <select id="toolbar-lang-select" style={{maxWidth: (this.state.langOn ? "10em" : "0")}} onChange={(e) => {I18N.fetchLanguageFile(e.target.value);}}>
            <option value="en_us">English</option>
            <option value="xx_pirate">Pirate Speak</option>
          </select>
        </div>
      </div>
    </div>;
  }
}

export default Toolbar;
