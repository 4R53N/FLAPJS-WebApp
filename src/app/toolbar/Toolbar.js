import React from 'react';
import HamburgerIcon from './HamburgerIcon.js';
import {
  Collapse,
  Navbar,
  NavbarToggler,
  NavbarBrand,
  Nav,
  NavItem,
  NavLink,
  UncontrolledDropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem } from 'reactstrap';

import './Toolbar.css';
import SaveIcon from './SaveIcon';
import UndoIcon from './UndoIcon';
import RedoIcon from './RedoIcon';
import NewIcon from './NewIcon';

class Toolbar extends React.Component
{
  constructor(props) {
    super(props);

    this.setType = this.setType.bind(this);
    this.state = {
      isOpen: false,
      type: "DFA"
    };
  }

  setType(curType) {
    this.setState({
      type: curType
    });
  }

  render()
  {
    const app = this.props.app;
    const drawer = document.getElementsByTagName('Drawer')[0];
    return(
        <div
          onClick={app.state.isOpen ?
            app.closeDrawer.bind(app) :
            app.openDrawer.bind(app)}>
          <Navbar className="navBar" dark expand="md">
            <NavbarBrand className="navTitle single-line" href="/" contenteditable="true">Untitled</NavbarBrand>
            <UncontrolledDropdown>
              <DropdownToggle className="navType" nav caret>
                {this.state.type}
              </DropdownToggle>
              <DropdownMenu left>
                <DropdownItem onClick={() => { this.setType("DFA"); }}>
                  DFA
                </DropdownItem>
                <DropdownItem onClick={() => { this.setType("NFA"); }}>
                  NFA
                </DropdownItem>
              </DropdownMenu>
            </UncontrolledDropdown>
            <Nav className="navOption" navbar>
              <NavItem className="navOption">
                <NavLink href="#">
                  <SaveIcon/>
                </NavLink>
              </NavItem>
              <UncontrolledDropdown nav inNavbar>
                <DropdownToggle nav caret>
                  <NewIcon/>
                </DropdownToggle>
                <DropdownMenu left>
                  <DropdownItem>
                    Unload File
                  </DropdownItem>
                  <DropdownItem>
                    Start From Scratch
                  </DropdownItem>
                </DropdownMenu>
              </UncontrolledDropdown>
            </Nav>
            <Nav className="navOption" navbar>
              <NavItem className="navOption">
                <NavLink href="#">
                  <UndoIcon/>
                </NavLink>
              </NavItem>
              <NavItem className="navOption">
                <NavLink href="#">
                  <RedoIcon/>
                </NavLink>
              </NavItem>
            </Nav>
            <HamburgerIcon active={app.state.isOpen}/>
          </Navbar>
        </div>
    );
  }
}

export default Toolbar;
