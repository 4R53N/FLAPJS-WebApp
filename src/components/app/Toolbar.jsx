import React from 'react';
import './Toolbar.css';

class Toolbar extends React.Component
{
    constructor(props)
    {
        super(props);
    }

    /** @override */
    render()
    {
        return (
            <div className="toolbar">
                <input type="text" value="what"/>
                <button>What?</button>
            </div>
        );
    }
}

export default Toolbar;
