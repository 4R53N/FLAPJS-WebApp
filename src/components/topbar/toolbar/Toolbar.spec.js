import React from 'react';
import { shallow } from 'enzyme';

import Toolbar from './Toolbar.jsx';

let wrapper;

beforeEach(() =>
{
    wrapper = shallow(<Toolbar/>);
});

describe('<Toolbar/> rendering', () =>
{
    test('renders correctly', () =>
    {
        expect(wrapper).toMatchSnapshot();
    });
});

describe('<Toolbar/> interaction', () =>
{
    /** I promise I will test you soon... */
});

describe('<Toolbar/> lifecycle', () =>
{
    /** ...later. */
});
