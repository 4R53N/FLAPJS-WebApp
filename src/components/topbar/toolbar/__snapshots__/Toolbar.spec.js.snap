// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toolbar/> rendering renders correctly 1`] = `
<div>
  <ToolbarDropdown
    list={
      Array [
        Object {
          "id": 0,
          "key": "file",
          "options": Array [
            Object {
              "id": 3,
              "key": "file/new",
              "selected": false,
              "title": "DFA",
            },
            Object {
              "id": 4,
              "key": "file/new",
              "selected": false,
              "title": "NFA",
            },
            Object {
              "id": 5,
              "key": "file/new",
              "selected": false,
              "title": "PDA",
            },
          ],
          "selected": false,
          "title": "New",
        },
        Object {
          "id": 1,
          "key": "file",
          "selected": false,
          "title": "Open",
        },
        Object {
          "id": 2,
          "key": "file",
          "selected": false,
          "title": "Save",
        },
      ]
    }
    title="File"
    toggleItem={[Function]}
  />
</div>
`;
